File: dfs_bfs/dfs_maze.py
# -*- coding: utf-8 -*-

"""
这里用dfs求解迷宫最短路径问题,问题和bfs的一样，只是求法不同
"""
directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

def dfs(maze, i, j, m, n, step, visit):
    if i < 0 or i > m-1 or j < 0 or j > n-1 \
            or visit[i][j] or maze[i][j] == 1:
        return
    step += 1
    visit[i][j] = True
    if i == m-1 and j == n-1:
        dfs.minimum = min(step, dfs.minimum)
    global directions
    for pos in directions:
        dfs(maze, i+pos[0], j+pos[1], m, n, step, visit)

if __name__ == '__main__':
    maze = [[0, 1, 0, 0, 0],
            [0, 0, 0, 1, 0],
            [0, 0, 0, 0, 0],
            [0, 1, 0, 1, 0],
            [0, 0, 0, 1, 0]]
    dfs.minimum = float('inf')
    m = len(maze)
    n = len(maze[0])
    visit = [[False for x in range(n)] for x in range(m)]
    dfs(maze, 0, 0, m, n, 0, visit)
    print(dfs.minimum)

File: dfs_bfs/dfs_maze.cpp
#include <iostream>
#include <vector>
#include <limits.h>
using namespace std;

class Solution{
    public:
        int minimum = INT_MAX;
        vector<vector<int>>matrix;
        vector<vector<int>>pos{{-1,0}, {0, 1}, {1, 0}, {0, -1}};
        void dfs(int m, int n, int i, int j, int step, vector<vector<bool>>visit);
        int find(vector<vector<int>>matrix);
};

void Solution::dfs(int m, int n, int i, int j, int step, vector<vector<bool>>visit){
    if (i < 0 || i > m || j < 0 || j > n || visit[i][j] || matrix[i][j] == 1){
        return;
    }
    //cout<<i<<","<<j<<" "<<m<<","<<n<<" "<<step<<endl;
    if (i == m && j == n){
        minimum = min(minimum, step);
        return;
    }
    visit[i][j] = true;
    for (int k=0; k<pos.size(); k++){
        dfs(m, n, i+pos[k][0], j+pos[k][1], step+1, visit);
    }
}

int Solution::find(vector<vector<int>>mat){
    matrix = mat;
    int m = mat.size();
    int n = mat[0].size();
    vector<vector<bool>>visit;
    for (int i=0; i<m; i++){
        vector<bool>tmp;
        for (int j=0; j<n; j++){
            tmp.push_back(false);
        }
        visit.push_back(tmp);
    }
    dfs(m-1, n-1, 0, 0, 0, visit);
    return minimum;
}
int main(void){
    vector<vector<int>> maze = {
        {0, 1, 0, 0, 0},
        {0, 0, 0, 1, 0},
        {0, 0, 0, 0, 0},
        {0, 1, 0, 1, 0},
        {0, 0, 0, 1, 0}
    };
    Solution s;
    int step = s.find(maze);
    cout<<"Minimum step: "<<step<<endl;
    return 0;
}
